confmap/expand.go

```diff
@@ -0,0 +1,197 @@
+	}
+
+	return uri
+}
+
+// findAndExpandURI attempts to find and expand the first occurrence of an expandable URI in input.
+func (mr *Resolver) findAndExpandURI(ctx context.Context, input string) (output string, changed bool, err error) {
+	var repl string
+	uri := findURI(input)
+	if uri != "" {
+		repl, changed, err = mr.expandURI(ctx, uri)
+		input = strings.ReplaceAll(input, uri, repl)
+	} else {
+		// Check if the next URI in input is expandable.
+		input, changed, err = mr.nextExpandableURI(ctx, input)
+	}
+	return input, changed, err
+}
+
+// nextExpandableURI attempts to find and expand the next expandable URI in input. nextExpandableURI is called
+// when the first occurrence of URI in input is not expandable.
+func (mr *Resolver) nextExpandableURI(ctx context.Context, input string) (string, bool, error) {
+	var err error
+	var expandedRemaining string
+	var changed bool
+
+	closeIndex := strings.Index(input, "}")
+	noExpand := input[:closeIndex+1]
+	remaining := input[closeIndex+1:]
+
+	// if remaining does not contain }, there are no URIs left: stop recursion.
+	if strings.Contains(remaining, "}") {
+		expandedRemaining, changed, err = mr.findAndExpandURI(ctx, remaining)
+		return noExpand + expandedRemaining, changed, err
+	}
+	return input, changed, err
+}
+
+// expandURI tries to expand uri. If an expandable URI is found, it returns the uri expanded, true and nil.
+// Otherwise, it returns the unchanged input, false and the expanding error.
+func (mr *Resolver) expandURI(ctx context.Context, uri string) (string, bool, error) {
+	expanded, changed, err := mr.expandStringURI(ctx, uri)
+	if err != nil {
+		return uri, changed, err
+	}
+	val := reflect.ValueOf(expanded)
+	switch val.Kind() {
+	case reflect.String:
+		return val.String(), changed, err
+	case reflect.Int, reflect.Int32, reflect.Int64:
+		return strconv.FormatInt(val.Int(), 10), changed, err
+	case reflect.Float32, reflect.Float64:
+		return strconv.FormatFloat(val.Float(), 'f', -1, 64), changed, err
+	case reflect.Bool:
+		return strconv.FormatBool(val.Bool()), changed, err
+	default:
+		return uri, changed, fmt.Errorf("expanding %v, expected string value type, got %T", uri, expanded)
+	}
+}
+
+func (mr *Resolver) expandStringURI(ctx context.Context, uri string) (any, bool, error) {
+	lURI, err := newLocation(uri[2 : len(uri)-1])
+	if err != nil {
+		return nil, false, err
+	}
+	if strings.Contains(lURI.opaqueValue, "$") {
+		return nil, false, fmt.Errorf("the uri %q contains unsupported characters ('$')", lURI.asString())
+	}
+	ret, err := mr.retrieveValue(ctx, lURI)
+	if err != nil {
+		return nil, false, err
+	}
+	mr.closers = append(mr.closers, ret.Close)
+	val, err := ret.AsRaw()
+	return val, true, err
+}
+
+type location struct {
+	scheme      string
+	opaqueValue string
+}
+
+func (c location) asString() string {
+	return c.scheme + ":" + c.opaqueValue
+}
+
+func newLocation(uri string) (location, error) {
+	submatches := uriRegexp.FindStringSubmatch(uri)
```

Comment:
I think uriRegexp is only used in this file, so move it here.

Reply:
Also moved other vars which were only used in this file.